### ISSUE TITLE
"Cannot select" LLVM failure due to unimplemented atomics on msp430

### ISSUE THREAD

**Issue by cr1901** [rust-lang/compiler-builtins#441]
# Context

A few nights ago, CI started [failing](https://github.com/cr1901/AT2XT/runs/4269789260?check_suite_focus=true#step:9:77) for some msp430 firmware I use to make sure that msp430 Rust code builds:

```
   Compiling rustc-std-workspace-core v1.99.0 (/usr/share/rust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/rustc-std-workspace-core)
LLVM ERROR: Cannot select: 0x7fbca0892c98: i16,ch = AtomicLoad<(load unordered (s16) from %ir.49)> 0x7fbca08313e8, 0x7fbca0893340
  0x7fbca0893340: i16,ch = CopyFromReg 0x7fbca08313e8, Register:i16 %19
    0x7fbca08a1c30: i16 = Register %19
In function: memcpy
error: could not compile `compiler_builtins`
```

MSP430 doesn't have meaningful atomic support, so I'm not sure why code using atomics is being generated.

# Bisecting `rustc`
Using [`cargo-bisect-rustc`](https://github.com/rust-lang/cargo-bisect-rustc), I found the range of commits where compilation started failing.

searched nightlies: from nightly-2021-11-18 to nightly-2021-11-19
regressed nightly: nightly-2021-11-19
searched commits: from https://github.com/rust-lang/rust/commit/c9c4b5d7276297679387189d96a952f2b760e7ad to https://github.com/rust-lang/rust/commit/cc946fcd326f7d85d4af096efdc73538622568e9
regressed commit: https://github.com/rust-lang/rust/commit/b6f580acc0ce233d5c4d1f9680d354fded88b824

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --preserve-target --preserve --start=2021-11-18 --end=2021-11-19 --with-src
```
</details>

I further bisected manually, the _actual_ last good commit in rust is https://github.com/rust-lang/rust/commit/a3b9405ae7bb6ab4e8103b414e75c44598a10fd2 and the regression is https://github.com/rust-lang/rust/commit/88f1bf73ee8466d07b1e38755de53a70fc292e20. Commit b6591c2 bumps `compiler_builtins` to `0.1.52`.

# Bisecting `compiler_builtins`

Using a nightly from before the failure started (see "Other Notes"), the first compile failure I get is with commit https://github.com/rust-lang/compiler-builtins/commit/ce86d41b4f95e4d1917460ee7a05fcd7fe6831fc. The error is quite similar to the one in the CI failures:

```
william@xubuntu-dtrain:~/Projects/toolchains/compiler-builtins$ cargo build --t[14/661]p430-none-elf -Zbuild-std=core
   Compiling compiler_builtins v0.1.49 (/home/william/Projects/toolchains/compiler-builtins)
warning: an associated function with this name may be added to the standard library in the future
 --> src/float/pow.rs:8:19
  |
8 |     let mut pow = i32::abs_diff(b, 0);
  |                   ^^^^^^^^^^^^^
  |
  = note: `#[warn(unstable_name_collisions)]` on by default
  = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
  = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
  = help: call with fully qualified syntax `Int::abs_diff(...)` to keep using the current method
  = help: add `#![feature(int_abs_diff)]` to the crate attributes to enable `num::<impl i32>::abs_diff`

LLVM ERROR: Cannot select: 0x7f6cd2888208: i16,ch = AtomicLoad<(load unordered (s16) from %ir.188)> 0x7f6cd287aa28:1, 0x7f6cd287aa28, src/mem/impls.rs:65:29 @[ src/mem/impls.rs:109:13 @[ src/mem/mod.rs:25:5 ] ]
  0x7f6cd287aa28: i16,ch = load<(dereferenceable load (s16) from %ir.28)> 0x7f6cd287b270, FrameIndex:i16<24>, undef:i16, src/mem/impls.rs:65:69 @[ src/mem/impls.rs:109:13 @[ src/mem/mod.rs:25:5 ] ]
    0x7f6cd287de38: i16 = FrameIndex<24>
    0x7f6cd28427b8: i16 = undef
In function: memcpy
warning: `compiler_builtins` (lib) generated 1 warning
error: could not compile `compiler_builtins`; 1 warning emitted
```

I believe fixing the failure in `compiler_builtins` will fix the error I'm seeing when compiling `rustc`, and I would need to test locally. That said, I'm not sure what is causing `compiler_builtins` to be built anyway when I specify `-Zbuild-std=core`. Doesn't `core` have _no_ dependencies?

# Solution

I'm not sure what should be done here. Can there be a gate for targets whose [maximum atomic width](https://github.com/rust-lang/rust/pull/55450/files#diff-ec783dc29738ed11a89d179827592876b42eab4bf7935adeff51aa54e05805bbR37-R43) is `0`? Although it eventually needs to be done, implementing atomic support for msp430 is unlikely right now, as I don't have the bandwidth [to do it](https://github.com/rust-lang/rust/issues/54511#issuecomment-433726311). Additionally, AIUI, Rust does not actually _mandate_ that targets implement atomics, and that they should become noops/not available on targets without them (not just msp430).

# Other Notes

### `rustc`
```
william@xubuntu-dtrain:~/Projects/toolchains/compiler-builtins$ rustc -Vv
rustc 1.58.0-nightly (ff0e14829 2021-10-31)
binary: rustc
commit-hash: ff0e14829e1806ca0d4226595f7fdf3e8658758f
commit-date: 2021-10-31
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

### `cargo`
```
william@xubuntu-dtrain:~/Projects/toolchains/compiler-builtins$ cargo -Vv
cargo 1.58.0-nightly (6c1bc24b8 2021-10-24)
release: 1.58.0
commit-hash: 6c1bc24b8b49d4bc965f67d7037906dc199c72b7
commit-date: 2021-10-24
host: x86_64-unknown-linux-gnu
libgit2: 1.3.0 (sys:0.13.23 vendored)
libcurl: 7.79.1-DEV (sys:0.4.49+curl-7.79.1 vendored ssl:OpenSSL/1.1.1l)
os: Ubuntu 20.04 (focal) [64-bit]
```

## Circular Dependency Problems
I have to compile on msp430 using `-Zbuild-std=core`, since it's not built as part of a Rust release right now. Even if I have an older version of `compiler_builtins` checked out, using a too-recent nightly/commit causes `-Zbuild-std=core` to compile `compiler_builtins` `v0.1.52` anyway as a dependency of the older `compiler_builtins` in a fun circular dependency!

```
william@xubuntu-dtrain:~/Projects/toolchains/compiler-builtins$ LD_LIBRARY_PATH[20/450]william/Projects/toolchains/build-llvm-toolchain/build-llvm/lib:/home/william/Projects/toolchains/build-llvm-toolchain/build-rust/build/x86_64-unknown-linux-gnu/stage1/lib" cargo +msp430-fix build --target=msp430-none-elf -Zbuild-std=core
   Compiling compiler_builtins v0.1.52
   Compiling core v0.0.0 (/home/william/Projects/toolchains/build-llvm-toolchain/build-rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/src/rust/library/core)
   Compiling compiler_builtins v0.1.49 (/home/william/Projects/toolchains/compiler-builtins)
   Compiling rustc-std-workspace-core v1.99.0 (/home/william/Projects/toolchains/build-llvm-toolchain/build-rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/src/rust/library/rustc-std-workspace-core)
warning: an associated function with this name may be added to the standard library in the future
 --> src/float/pow.rs:8:19
  |
8 |     let mut pow = i32::abs_diff(b, 0);
  |                   ^^^^^^^^^^^^^
  |
  = note: `#[warn(unstable_name_collisions)]` on by default
  = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
  = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
  = help: call with fully qualified syntax `Int::abs_diff(...)` to keep using the current method
  = help: add `#![feature(int_abs_diff)]` to the crate attributes to enable `num::<impl i32>::abs_diff`

LLVM ERROR: Cannot select: t14: i16,ch = AtomicLoad<(load unordered (s16) from %ir.188)> t13:1, t13, /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/impls.rs:65:29 @[ /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/impls.rs:109:13 @[ /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/mod.rs:25:5 ] ]
  t13: i16,ch = load<(dereferenceable load (s16) from %ir.28)> t12, FrameIndex:i16<24>, undef:i16, /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/impls.rs:65:69 @[ /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/impls.rs:109:13 @[ /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/mod.rs:25:5 ] ]
    t11: i16 = FrameIndex<24>
    t5: i16 = undef
In function: memcpy
error: could not compile `compiler_builtins`
warning: build failed, waiting for other jobs to finish...
warning: `compiler_builtins` (lib) generated 1 warning
error: build failed
```

**Comment by Amanieu** [comment]
I think the easiest way to resolve this is to make a minor change to `ce86d41b4f95e4d1917460ee7a05fcd7fe6831fc`: use a volatile load (`ptr::read_volatile`) instead of `atomic_load_unordered`. This can be made conditional on `cfg(target_has_atomic_load_store = "ptr")` so that the change only applies to msp430.

**Comment by cr1901** [comment]
@Amanieu I'll prepare a patch for this/try it out and open a PR. Eventually, the need for this for msp430 will be removed; I've asked someone familiar w/ LLVM to help me implement fences for msp430 (because I am empathetically, err, not familiar with LLVM :)). But it's prob good to gate it for others.

**Comment by ducktec** [comment]
I'm believe I see the [same issue](https://github.com/ducktec/esp32c3-hal/runs/4291792834?check_suite_focus=true#step:4:105) in the nightly channel with the ESP32-C3 and the `rv32imc` ISA. This ISA does not support atomic load/store operations.

```bash
   = note: rust-lld: error: undefined symbol: __atomic_load_4
          >>> referenced by impls.rs:98 (/cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.52/src/mem/impls.rs:98)
          >>>               compiler_builtins-f61960a02225fda8.compiler_builtins.f451d05e-cgu.3.rcgu.o:(memcpy) in archive /usr/share/rust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/riscv32imc-unknown-none-elf/lib/libcompiler_builtins-f61960a02225fda8.rlib
```
So I guess that the alternative (non-atomic) load would have to be made conditional not only for the msp430, but at least also for the RISCV ISA without atomic support. And for those, the changes will have to stick around.

I tried replacing the atomic operation with `core::ptr::read_volatile` and with that change the issue was no longer observable.

**Comment by bjorn3** [comment]
Doesn't riscv require loads and stores to be atomic even when the atomic extension isn't implemented? From section 2.6 of the riscv spec:

> Furthermore, whereas **naturally aligned loads and stores are guaranteed to execute atomically**, misaligned loads and stores might not, and hence require additional synchronization to ensure atomicity.

(emphasis mine)

The loads should be aligned and thus it should be fine to depend on the atomicity of aligned loads. In addition the fence instruction is also part of the base spec. Not that it is needed for this specific case as the unordered memory ordering is used.

**Comment by Amanieu** [comment]
IIRC there were concerns about the compatibility between a direct instruction for load/store and libcalls for RMW operations. So LLVM just makes everything use libcalls on RISCV without A. In any case, the workaround I described (`cfg(target_has_atomic_load_store = "ptr")`) should also apply here.

**Comment by cr1901** [comment]
I'll make sure to add a comment describing the RISCV case as well. But just to clarify for my own understanding:

> In any case, the workaround I described (`cfg(target_has_atomic_load_store = "ptr")`) should also apply here.

If RISCV without A is preferring a libcall, shouldn't the atomic load be converted into a libcall _without_ bombing as above? Meaning either version of the load (`cfg(target_has_atomic_load_store = "ptr")` or `cfg(not(target_has_atomic_load_store = "ptr"))`) should work?

>Not that it is needed for this specific case as the unordered memory ordering is used.

Why does Rust generate a _compiler_ fence in the output LLVM when attempting to lower an _unordered_ atomic load (which is the crux of the problem in the msp430 case- there is no ?

**Comment by bjorn3** [comment]
> Why is an atomic load w/ unordered memory ordering generating a fence at all, software or otherwise, which I assume is the actual cause of the "cannot select AtomicLoad"?

It doesn't try to generate a fence. It tries to generate an atomic load. Fences are only necessary to ensure a consistent view of the memory. Unordered atomic operations don't generate fences. They only need an instruction that atomically performs the operation, without needing any memory ordering restrictions. As the msp430 is single core only AFAICT, it should be possible to lower atomic loads and stores to regular loads and stores inside LLVM. It is probably that simply nobody implemented this.

**Comment by cr1901** [comment]
>As the msp430 is single core only AFAICT, it should be possible to lower atomic loads and stores to regular loads and stores inside LLVM.

[Upstream](https://github.com/rust-lang/rust/pull/55450#issuecomment-433729789) seems to want `singlethreaded: true` to be enabled, which would require compiler fence support (just checked- the `singlethreaded` `bool` is still present). This feels in contrast to `singlethreaded: false` and lowering atomic loads/stores to regular loads/stores behind the scenes?

Anyways, I'm fine with the `cfg(target_has_atomic_load_store = "ptr")` solution. I just don't want it added _solely_ for my sake.

**Comment by asl** [comment]
MSP430 does not support anything "atomic" and it's up to the IR producer to ensure that the IR is compatible with the target hardware. While it might look reasonable to expect that backend would lower unsupported stuff to anything else (e.g. libcalls or just drop atomic from the load), in practice it is not: we cannot expect that e.g. PowerPC or RISC-V backend would emulate x86 vector intrinsics :)

**Comment by cr1901** [comment]
>and it's up to the IR producer to ensure that the IR is compatible with the target hardware

@asl Rust seems to think this is the [backend's](https://github.com/rust-lang/rust/pull/55450#issuecomment-433729789) responsibility. I'm not making a value judgment either way- I just want my msp430 code to work :D. Rust maintainers in the linked issue want the msp430 backend to support compiler fences. Which would also be a solution to this issue here.

Would you accept compiler fence support in the msp430 LLVM backend<sup>1</sup>? If not, I'll continue to submit issues and PRs to gate msp430 behavior like this one (and get clarification from Rust upstream).

1. Of course I'll submit this through Phabricator. I'm just asking here while you're around :P.

**Comment by asl** [comment]
> @asl Rust seems to think it is the backend's responsibility to accept certain IR regardless of arch- and atomic support is one of them. singlethreaded: true is the fallback for producing "IR compatible with the target hardware" in the case of msp430, but even this requires compiler fences. . Rust maintainers in the linked issue want the msp430 backend to support compiler fences. Which would also be a solution to this issue here.

Well let me sincerely disagree with this. And let me amplify the issue we're talking here. Consider the frontend will generate, say, ppc vector intrinsic. Or i256 division, or something else, because why not. Should it be a backend responsibility to take care of this? The common (unwritten contract) is no. It's up to frontend to generate the IR that is compatible with the given target (similar to ABI bits, btw).

Certainly the key point here is where to draw the borderline. If this is only simple thing that just drops some bits and lowers atomic stuff to normal loads / stores, then probably it could be fine. I just do not want some generalizations and precedents :)

**Comment by bjorn3** [comment]
> Consider the frontend will generate, say, ppc vector intrinsic.

No, it shouldn't handle ppc vector intrinsics as those are powerpc specific, but it should handle the platform independent vector intrinsics implemented by every other backend. In fact we recently got `core::simd` which uses the platform independent vector intrinsics on all platforms, even those that don't natively support simd. For those platforms the backend is responsible for lowering it to scalar instructions.

**Comment by cr1901** [comment]
>For those platforms the backend is responsible for lowering it to scalar instructions.

Almost certainly msp430 doesn't support these either, but you shouldn't be doing simd on an msp430 ;).

I still think it's on topic for this issue to discuss what to do re: msp430 here because there's a few conflicting things going on now:

1. I want to eventually get msp430 support stable. In the interim, I want to play nice w/ upstream Rust, so they don't start thinking "msp430 has too many exceptions catering to it". 
    1. For instance: for a while, msp430 was the only backend using the `no_integrated_as` `Target` option, until it got integrated LLVM assembler support. At that point `no_integrated_as` was [removed](https://github.com/rust-lang/rust/pull/70345#issuecomment-849523632), to possibly be readded in a more constrained form.
2. To remove exceptions catering to msp430, Upstream Rust wants the missing msp430 support to be implemented one way.
3. @asl, the backend designer/maintainer working on the C side of things (I know you wrote the backend, are you still `CODE_OWNER`?), wants to maintain the msp430 backend in a different way.

Adding `cfg(target_has_atomic_load_store = "ptr")`, though it's the easiest solution, is another exception catering to msp430. _I_ personally don't mind, but I'd like to avoid giving upstream reasons to consider removing msp430 :D.

**Comment by asl** [comment]
Well, let me be clear: I'm not opposing adding few emulations here and there. But I do not see point to support every and each IR feature on msp430 :)

**Comment by ducktec** [comment]
Regarding the situation in `rv32imc`, I have to admit that I'm a bit out of my depth here. I got this error in my CI builds and had the intention to report it when seeing that it somewhat lines up with the MSP430 issue.

This is what I understand the situation to be (please correct me if I got something wrong):
- In https://github.com/rust-lang/rust/pull/53822#issuecomment-417420859 it was found for `rv32imc` that:
  >  tl;dr can't emit fence-based inline sequences for some operations and libcalls for other operations on the same   memory location because those two might not synchronize correctly with each other.

  So, libcalls it is, even though I understand that from an ISA specification perspective converting this specific case to a regular load (without fencing) would be sound. 
  With the specification of `max_atomic_width: Some(0)` it seems not possible to further detail that some atomic operations are feasible without fencing and thus do not contradict the tl;dr quoted above. It's just either all atomic load/store operations (including the ordered ones requiring fencing) or none at all, correct?
- And from https://github.com/rust-lang/rust/pull/66548 I take that the libcall to `__atomic_load_4` should also not take place at all since it would require linking against some additional implementation that provides this?

So, if `core::intrinsics::atomic_load_unordered` inevitably leads to a libcall to `__atomic_load_4` which is considered not acceptable, I understand it is necessary to make these statements conditional as proposed. Positive side effect: introducing this could not be blamed on msp430 ;).

**Comment by workingjubilee** [comment]
> > For those platforms the backend is responsible for lowering it to scalar instructions.
> 
> Almost certainly msp430 doesn't support these either, but you shouldn't be doing simd on an msp430 ;).
> 

People are not just using `core::simd` to twist pixels on consumer PCs or run high-power data processing on big iron, they are using `core::simd` as a high level API for expressing digital signal algorithms that are appropriate for low-power microcontrollers as well. I am of the understanding Texas Instruments even offers their own software library for this purpose.

I hope to offer alternate lowerings as a fallback, to make it easier for backends and reduce the maintenance burden, but as the generic vector ops functionally just express operations over an array-like construct (with some qualifications so that it can be easily optimized if the backend knows how), I don't see why, in principle, such IR cannot be lowered to small devices.

**Comment by cr1901** [comment]
@workingjubilee I don't maintain the LLVM backend (nor am I terribly familiar w/ it, tbh), and while that's something I need to/been meaning to fix, I'm the only person right now making sure Rust on msp430 works. I simply do not have the bandwidth to learn how to add all the unimplemented IR constructs, only for my contributions to potentially get rejected by upstream.

>I'm not opposing adding few emulations here and there.

Although, based on above, if the only missing IR constructs are compiler fences and SIMD, then I can probably get away w/ saying "this is required for stabilizing Rust support, and this is the only reason I'm implementing it."

**Comment by workingjubilee** [comment]
Hmm. If getting it upstreamed is a concern, then we can probably figure something out, either adjusting how we generate the IR in the first place, or even having LLVM run a "desugaring" pass for vector LLVMIR using the pass manager.

**Comment by awygle** [comment]
> Well, let me be clear: I'm not opposing adding few emulations here and there. But I do not see point to support every and each IR feature on msp430 :)

Would you be open to supporting AtomicFence on MSP430 in the same way as [the AVR backend supports it](https://github.com/llvm/llvm-project/blob/e2d8b607427d7d01bd9bdc007af4f4a4ac9833c6/llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp#L1025)? This should allow us to [use singlethread: true](https://github.com/rust-lang/rust/issues/54511#issuecomment-433726311) for MSP430 in Rust.

If that sounds acceptable, I'd be happy to help with implementing it.

**Comment by cr1901** [comment]
cc: @asl What are your thoughts re: SIMD and AtomicFence? Are you open to those emulations? I _think_ these are the only emulations required so that I can get `libcore` for msp430 enabled for building in CI (which is a big step to getting msp430 stabilized).
